storagemarket/impl/shardregistration.go

```diff
@@ -0,0 +1,146 @@
+package storageimpl
+
+import (
+	"context"
+
+	"github.com/ipfs/go-datastore"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/dagstore"
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/lotus/chain/actors/builtin/miner"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/extern/sector-storage/storiface"
+	"github.com/filecoin-project/specs-storage/storage"
+
+	mktdagstore "github.com/filecoin-project/go-fil-markets/dagstore"
+	"github.com/filecoin-project/go-fil-markets/storagemarket"
+	"github.com/filecoin-project/go-fil-markets/storagemarket/impl/providerstates"
+)
+
+var shardRegKey = datastore.NewKey("shards-registered")
+
+type SectorState interface {
+	StateSectorGetInfo(context.Context, address.Address, abi.SectorNumber, types.TipSetKey) (*miner.SectorOnChainInfo, error)
+	IsUnsealed(ctx context.Context, sector storage.SectorRef, offset storiface.UnpaddedByteIndex, size abi.UnpaddedPieceSize) (bool, error)
+}
+
+type ShardRegistration struct {
+	maddr       address.Address
+	ds          datastore.Datastore
+	dagStore    mktdagstore.DagStoreWrapper
+	sectorState SectorState
+}
+
+func NewShardRegistration(
+	maddr address.Address,
+	ds datastore.Datastore,
+	dagStore mktdagstore.DagStoreWrapper,
+	sectorState SectorState,
+) *ShardRegistration {
+	return &ShardRegistration{
+		maddr:       maddr,
+		ds:          ds,
+		dagStore:    dagStore,
+		sectorState: sectorState,
+	}
+}
+
+func (r *ShardRegistration) registerShards(ctx context.Context, deals []storagemarket.MinerDeal) error {
+	// Check if all deals have already been registered as shards
+	has, err := r.ds.Has(shardRegKey)
+	if err != nil {
+		return xerrors.Errorf("failed to get shard registration status: %w", err)
+	}
+	if has {
+		// All deals have been registered as shards, bail out
+		return nil
+	}
+
+	// Filter for deals that are currently sealing.
+	// If the deal has not yet been handed off to the sealing subsystem, we
+	// don't need to call RegisterShard in this migration; RegisterShard will
+	// be called in the new code once the deal reaches the state where it's
+	// handed off to the sealing subsystem.
+	resch := make(chan dagstore.ShardResult, len(deals))
+	for _, deal := range deals {
+		if deal.Ref.PieceCid == nil {
+			continue
+		}
+
+		// Check if the deal has been handed off to the sealing subsystem
+		var sealing bool
+		for _, state := range providerstates.ProviderSealingStates {
```

Comment:
I think the `sealing` term is a bit of a misnomer here, because this also includes deals that have finished sealing.

Reply:
A more efficient way of doing this would be to form a `map[fsm.StateKey]struct{}` before the loop, and use an "exists" state to evaluate if the deal is in one of the required states.

Reply:
Changed `sealing` to `inSealingSubsystem`

It may be more efficient to use a map, however there are only 5 `ProviderSealingStates` and this is a one-off migration, so I think it's probably better to keep it simple at the cost of a few more milliseconds.

Reply:
The double iteration is ugly :-( set membership check is in Go is very idiomatic as a map. If you don't want to form a map here, then just have the original set be a map?